src/canister/user_index/src/api/canister_management/mod.rs

```diff
@@ -1,107 +1,233 @@
+        .iter()
+        .filter_map(|r| r.as_ref().ok().cloned())
+        .collect::<Vec<Principal>>();
 
+    let num_success = success_canisters.len();
 
-    run_task_concurrently(canister_reinstall_futures, 10, result_callback, || false).await;
+    CANISTER_DATA.with_borrow_mut(|canister_data| {
+        canister_data
+            .available_canisters
+            .extend(success_canisters.clone());
+
+        // remove the canisters that are recycled from user_principal_id_to_canister_id_map and unique_user_name_to_user_principal_id_map
+        // canister_id is the value in the map
+        for canister_id in success_canisters.clone() {
+            canister_data
+                .user_principal_id_to_canister_id_map
+                .retain(|_, v| *v != canister_id);
+            canister_data
+                .unique_user_name_to_user_principal_id_map
+                .retain(|_, v| *v != canister_id);
+        }
+    });
+
+    let failed_list = results
+        .iter()
+        .filter(|r| r.is_err())
+        .map(|r| r.as_ref().unwrap_err().clone())
+        .collect::<Vec<(Principal, String)>>();
 
-    Ok(format!("Sucess {}", canisters[0].to_string()))
-}
\ No newline at end of file
+    let end = get_current_system_time();
+
+    CANISTER_DATA.with_borrow_mut(|canister_data| {
+        canister_data.recycle_status = RecycleStatus {
+            success_canisters: success_canisters.iter().map(|c| c.to_string()).collect(),
+            last_recycled_at: Some(end),
+            last_recycled_duration: Some(end.duration_since(start).unwrap().as_secs()),
+            num_last_recycled_canisters: num_success as u64,
+            failed_recycling: failed_list,
+        };
+    });
+}
+
+pub async fn reset_canister(
+    canister_id: Principal,
+    individual_user_template_canister_wasm: CanisterWasm,
+) -> Result<Principal, (Principal, String)> {
+    // secondary check for not allowing registered canisters to be recycled
+    let (session_type_res,): (Result<SessionType, String>,) =
+        match ic_cdk::call(canister_id, "get_session_type", ()).await {
+            Ok(r) => r,
+            Err(e) => return Err((canister_id, e.1)),
+        };
+    if let Err(e) = session_type_res {
+        return Err((canister_id, e));
+    }
+    if session_type_res.unwrap() != SessionType::AnonymousSession {
+        return Err((
+            canister_id,
+            "Canister is not AnonymousSession. Can not recycle".to_string(),
+        ));
+    }
+
+    canister_management::recharge_canister_if_below_threshold(&canister_id)
+        .await
+        .map_err(|e| (canister_id, e))?;
+
+    // reinstall wasm
+    match reinstall_canister_wasm(
+        canister_id,
+        None,
+        individual_user_template_canister_wasm.version.clone(),
+        individual_user_template_canister_wasm.wasm_blob.clone(),
+    )
+    .await
+    {
+        Ok(_) => Ok(canister_id),
+        Err(e) => Err((canister_id, e)),
+    }
```

Comment:
After recycling I think we should remove all the additional cycles that are on top of the base 0.2T cycles?

Reply:
Any existing method I can use ?